src/ol/source/Vector.js

```diff
@@ -299,12 +299,12 @@ class VectorSource extends Source {
   addFeatureInternal(feature) {
     const featureKey = getUid(feature);
 
+    this.setupChangeEvents_(featureKey, feature);
+
     if (!this.addToIndex_(featureKey, feature)) {
       return;
```

Comment:
It would be better to remove the feature from the collection here, if there is a collection.

Reply:
Problem is, `addFeatureInternal` can be called from a collection `add` event, so it would mean removing an element while adding it... would that be possible?

See: https://github.com/openlayers/openlayers/blob/ddb7da37336f803c635d3becfd316047615944a6/src/ol/source/Vector.js#L451

Reply:
That should work.